### PR TITLE
Legacy GLES extensions.

### PR DESCRIPTION
**Pull request by luboslenco**
Added GL_EXT_draw_buffers, GL_OES_standard_derivatives and GL_EXT_shader_texture_lod extensions for legacy GLES. Let me know if I can improve further.

### CODE REVIEW

File: spirv_glsl.cpp [hunk 1]
```diff
@@ -1059,6 +1059,9 @@ void CompilerGLSL::replace_fragment_output(SPIRVariable &var)
 		if (location != 0)
 			throw CompilerError("Arrayed output variable used, but location is not 0. "
 			                    "This is unimplemented in SPIRV-Cross.");
+		
+		if (is_legacy_es())
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: What are the rules for this extension being needed? Is it needed to expose gl_FragData itself, or just FragData[N] where N > 0?

File: spirv_glsl.cpp [hunk 2]
```diff
@@ -4481,7 +4506,7 @@ string CompilerGLSL::image_type_glsl(const SPIRType &type)
 	if (type.image.arrayed)
 	{
 		if (!options.es && options.version < 130)
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Can probably add a is_legacy_desktop() here as well?

File: spirv_glsl.hpp [hunk 3]
```diff
@@ -337,6 +337,11 @@ class CompilerGLSL : public Compiler
 		return (options.es && options.version < 300) || (!options.es && options.version < 130);
 	}
 
+	inline bool is_legacy_es() const
+	{
+		return (options.es && options.version < 300);
```
Comment (HansKristian-ARM) [inline spirv_glsl.hpp]: No need for the ().

File: spirv_glsl.cpp (hunk 1 above)
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: In that case, wouldn't this be needed in non-arrayed output as well, where location != 0?

File: spirv_glsl.cpp [hunk 4]
```diff
@@ -1823,11 +1826,21 @@ string CompilerGLSL::legacy_tex_op(const std::string &op, const SPIRType &imgtyp
 	if (op == "texture")
 		return join("texture", type);
 	else if (op == "textureLod")
-		return join("texture", type, "Lod");
+	{
+		if (is_legacy_es())
+			return join("texture", type, "LodEXT");
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Seems cleaner with a ternary here.

File: spirv_glsl.cpp [hunk 5]
```diff
@@ -1823,11 +1826,21 @@ string CompilerGLSL::legacy_tex_op(const std::string &op, const SPIRType &imgtyp
...
+		else 
+			return join("texture", type, "Lod");
+	}
 	else if (op == "textureProj")
 		return join("texture", type, "Proj");
 	else if (op == "textureProjLod")
-		return join("texture", type, "ProjLod");
+	{
+		if (is_legacy_es())
+			return join("texture", type, "ProjLodEXT");
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Simpler with a ternary.

File: spirv_glsl.cpp [hunk 6]
```diff
@@ -3894,14 +3913,20 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
...
 	case OpImageSampleExplicitLod:
-	case OpImageSampleProjImplicitLod:
 	case OpImageSampleProjExplicitLod:
-	case OpImageSampleDrefImplicitLod:
 	case OpImageSampleDrefExplicitLod:
-	case OpImageSampleProjDrefImplicitLod:
 	case OpImageSampleProjDrefExplicitLod:
+	{
+		emit_texture_op(instruction);
+		if (is_legacy_es())
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: I would move this test inside emit_texture_op. It is virtual and it doesn't make sense to run this for C++/Metal backends.

File: spirv_glsl.cpp [hunk 7]
```diff
@@ -1820,14 +1826,17 @@ string CompilerGLSL::legacy_tex_op(const std::string &op, const SPIRType &imgtyp
 		break;
 	}
 
+	if (is_legacy_es() && (op == "textureLod" || op == "textureProjLod"))
+		require_extension("GL_EXT_shader_texture_lod");
```
Comment (luboslenco) [inline spirv_glsl.cpp]: Maybe like this?

**Comment by luboslenco** [comment]
Changes pushed.

File: spirv_glsl.cpp [hunk 8]
```diff
@@ -1820,14 +1826,17 @@ string CompilerGLSL::legacy_tex_op(const std::string &op, const SPIRType &imgtyp
...
+
 	if (op == "texture")
 		return join("texture", type);
 	else if (op == "textureLod")
-		return join("texture", type, "Lod");
+		return is_legacy_es() ? join("texture", type, "LodEXT") : join("texture", type, "Lod");
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: join("texture", type, is_legacy_es() ? "LodEXT" : "Lod") is cleaner.

Comment (luboslenco) [inline spirv_glsl.cpp]: How did I miss that.. Pushed now.

**Pull request merged**